WIP: nbgrader authentication

jdfreder opened the pull request:
Hey @jhamrick ,

Here's a simple extensible authentication scheme.  

FormgradeApp.authenticator_class can be used to change the authenticator.  `BaseAuth` is the base authenticator class.  A `NoAuth` class is provided as the default, and retains current behavior of `formgrade`.  I moved the `start_nbserver` traitlet and nbserver launching logic into `NoAuth`.

The `HubAuth`, empty now, will contain the logic for authenticating with JupyterHub.

Let me know what you think of this design.  I've tested it locally and everything seems to work as-is.  The not yet implement `HubAuth` can be tested quickly (when it's done) using `nbgrader formgrade --FormgradeApp.authenticator_class='nbgrader.auth.hubauth.HubAuth'`.

closes #145

jhamrick commented on nbgrader/html/formgrade.py:
```diff
@@ -169,20 +205,19 @@ def view_submission(submission_id):
-            os.path.relpath(filename, app.notebook_dir))
+    if server_exists:
+        relative_path = os.path.relpath(filename, app.notebook_dir)
+        resources['notebook_path'] = app.auth.get_notebool_url(relative_path)
```
I think this should be `get_notebook_url`

jhamrick commented on requirements.txt:
```diff
@@ -1,4 +1,3 @@
-ipython==3.1.0.dev0
```
Is there a reason for removing this (rather than I guess changing the version number)?

jdfreder replied:
No that's a Jon install hack, I accidentally checked it in.

jhamrick commented:
This looks awesome -- thanks for working on this! I think Travis is unhappy because you need to add some stuff to the `__init__.py` for auth, and I left a few small comments inline. Overall though I think this design works really well!

jdfreder commented:
@jhamrick @minrk @ellisonbg this now works with JupyterHub auth.  I had to convert all of the URLs in the templates to relative paths, since the JHub auth nests the entire blueprint (we are using a blueprint now instead of registering routes directly on the flask app) in a sub path `hub/formgrade`.  @jhamrick I still haven't verified that all of the URLs map correctly, but most of them do.  Any path that's more than one level deep probably won't map correctly since I just preprended `..`  to the URLs.

To test, without writing any config files, run the following:
```bash
export CONFIGPROXY_AUTH_TOKEN=alice; jupyterhub
```

and in another terminal:

```bash
export JPY_API_TOKEN=`jupyterhub token`; export CONFIGPROXY_AUTH_TOKEN=alice; nbgrader formgrade --FormgradeApp.authenticator_class='nbgrader.auth.hubauth.HubAuth' --HubAuth.graders='["jon"]'
```

I want to do some cleanup, by using config to pass in the API token.

jdfreder commented:
Hehe, I'm not sure how coverage increased.  I didn't write a single test!

jhamrick commented:
Awesome, I will test this out later tonight or tomorrow. I am also working on some basic tests for the formgrader that just check that all the links work properly, so that should be perfect for verifying that everything still works -- I'll finish up with that soon and merge it so then we can test this PR against it.

minrk commented on nbgrader/apps/formgradeapp.py:
```diff
@@ -54,8 +47,11 @@ class FormgradeApp(BaseNbGraderApp):
-    start_nbserver = Bool(True, config=True, help="Start a single notebook server that allows submissions to be viewed.")
+    directory_format = Unicode('{notebook_id}.ipynb', config=True, help="""Format
+        string for the directory structure of the autograded notebooks""")
+    authenticator_class = DottedObjectName('nbgrader.auth.noauth.NoAuth', config=True, help="""
```
you can use `authenticator_class = Type('nbgrader.auth...')` and avoid needing the DottedObjectName + import_string.

jdfreder replied:
Nifty!

minrk commented on nbgrader/auth/hubauth.py:
```diff
@@ -0,0 +1,123 @@
+
+    proxy_address = Unicode(None, config=True, allow_none=True, help="Address of the configurable-http-proxy server.")
+    proxy_port = Int(8001, config=True, help="Port of the configurable-http-proxy server.")
+    hub_address = Unicode(None, config=True, allow_none=True, help="Address of the hub server.")
```
These addresses don't need to allow None, they can have 

```python
def _hub_address_default(self):
    return self._ip
```

No need for the init logic below.

minrk commented on nbgrader/auth/hubauth.py:
```diff
@@ -0,0 +1,123 @@
+        return "http://{}:{}/notebooks/{}".format(
+            self.hub_address,
+            self.hub_port,
+            relative_path)
```
This isn't the right URL, is it? Shouldn't it be the public (proxied) URL of the user's server?

jdfreder replied:
Ah yes, you're right.  It should be /user/notebooks.  I haven't tested this with notebooks yet, but will test today.

minrk commented on nbgrader/html/templates/notebook_submissions.tpl:
```diff
@@ -1,8 +1,8 @@
 {%- block breadcrumb -%}
-<li><a href="/assignments">Assignments</a></li>
-<li><a href="/assignments/{{ assignment_id }}">{{ assignment_id }}</a></li>
+<li><a href="../assignments">Assignments</a></li>
```
Is this jinja? If so, a `{{base_url}}` might be more robust than `..` everywhere.

jdfreder replied:
Yeah, it is, and yes that's a good suggestion.

jhamrick commented:
@jdfreder #164 are the tests that I've been working on -- they're just about ready to merge, except that they keep hanging on travis after completing. I'll try to figure out why tomorrow and let you know once I've actually merged them. It's possible there might be a merge conflict with this (I had to add just a few things to `formgradeapp.py`) but probably nothing too major.

jhamrick commented:
Ok, #164 has been merged now!

jdfreder commented:
Thanks for the review @minrk and thanks for letting me know @jhamrick .  I'll rebase today and try testing the actual complete grading process.

jhamrick commented:
I think if you add `nbgrader.auth` to the list of packages in `setup.py` that will at least fix your errors with the docs.

jdfreder commented:
Thanks

jdfreder commented:
@jhamrick tests are passing now.  Commit d62ec2c
relys on the new JHub cookie auth API https://github.com/jupyter/jupyterhub/pull/207 which hasn't been merged yet.  Until it's merged, I suggest testing by checking out f122d7d .

jhamrick commented:
Awesome, I will test this out later today after the meeting!

jhamrick commented:
This is looking really awesome. Some comments as I'm going through:

1. Currently, unless the notebooks and formgrader are all rooted at $HOME, opening up the live notebook won't work. Maybe it would be worth it to add a config option to the `HubAuth` class that is a prefix for the path to the notebooks that the formgrader is trying to access? Then `get_notebook_url` would add that prefix to the beginning of `relative_path`.

2. It seems like the css/js/fonts resources aren't being loaded when using `HubAuth` (seems to be fine for `NoAuth`, though). Looks like they are trying to be loaded from e.g. `/hub/static/css/bootstrap.min.css`, but probably this should be `/hub/formgrade/static/css/bootstrap.min.css`?

3. Similarly, the breadcrumbs at the top of the formgrader page (the one rendered from `formgrade.tpl`) seem to have invalid links -- e.g. `/assignments` rather than `/hub/formgrade/assignments`.

I wonder how difficult it would be to add another set of tests that also launch JupyterHub and then verify that all the links are correct there, too... I feel like probably it would not be too difficult. If you want I can try to look into this today.

jdfreder commented:
> Currently, unless the notebooks and formgrader are all rooted at $HOME, opening up the live notebook won't work. Maybe it would be worth it to add a config option to the HubAuth class that is a prefix for the path to the notebooks that the formgrader is trying to access? Then get_notebook_url would add that prefix to the beginning of relative_path.

Good idea!  Yeah, I've been running both servers from ~/ so I'm not surprised a bug like that exists.



> It seems like the css/js/fonts resources aren't being loaded when using HubAuth (seems to be fine for NoAuth, though). Looks like they are trying to be loaded from e.g. /hub/static/css/bootstrap.min.css, but probably this should be /hub/formgrade/static/css/bootstrap.min.css?

Hmmm, strange.  I think the `{{base_url}}` template resource that @minrk suggested fixed it for me.  Again, there must be assumptions leaking into the implementation about my install&execution (paths)- I'll investigate.

> seem to have invalid links

Sounds like {{base_url}} is missing, I'll look.

Thanks again for testing this, it helps me a lot.  The regression tests are probably a good idea, and worth the price of admission- having Travis install JHub and friends (node, npm, etc.) and if you're willing to look into it, that'd be great!  In the meantime I'll resolve the multi-path problem.

jhamrick commented:
Ok @jdfreder I made a pull request against your branch for the tests!

jdfreder commented:
Ok @jhamrick , I think I addressed all of the URL problems.  The biggest problem was a missing `resources.` prefix on `base_url` for the formgrade.tpl (since it's going through nbconvert).  I also added a `notebook_url_prefix` configurable that allows you to set the relative path of formgrader to JHub.  i.e. if JHub is started in `~/` and formgrader in `~/Documents/tests`, `notebook_url_prefix="Documents/tests"`.

I merged your tests, thanks!  Let's see if all is good and sound..

jhamrick commented:
Awesome, I am also just about finished writing some tests that will check that the url prefix works, so I'll run those against your changes and submit another PR against your branch soon.

jhamrick commented:
Oops, I broke the tests on python 2. This diff should fix it, I think:

```
diff --git a/nbgrader/tests/test_nbgrader_formgrade.py b/nbgrader/tests/test_nbgrader_for
index ad98948..1489164 100644
--- a/nbgrader/tests/test_nbgrader_formgrade.py
+++ b/nbgrader/tests/test_nbgrader_formgrade.py
@@ -10,7 +10,8 @@ from nose.tools import assert_equal
 from textwrap import dedent

 try:
-    from urllib import urljoin, unquote # Python 2
+    from urllib import unquote # Python 2
+    from urlparse import urljoin
 except ImportError:
     from urllib.parse import urljoin, unquote # Python 3
```

jhamrick commented:
And as for the python 3 tests, I think that line that just does `npm install` in travis is the problem. I copied that from jupyterhub, but probably shouldn't have.

jhamrick commented:
I am working on fixing travis stuff on the `auth` branch, I'll make a PR to your branch when I'm done!

jdfreder commented:
Woohoo!  Looks like your new tests pass @jhamrick .

jhamrick commented:
Yay! Thanks for all the work you've done on this @jdfreder , nice job.
